sdk/core/azure-core/src/main/java/com/azure/android/core/internal/util/CancellationTokenImpl.java

```diff
@@ -0,0 +1,326 @@
+     * is already in the cancelled state then the provided CallBack will be immediately executed.
+     *
+     * This is O(1) in best case and O(k) in worst, where k is number of concurrent threads in case
+     * of race.
+     *
+     * @param onCancel The onCancel Callback.
+     */
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    public void registerOnCancel(@NonNull Runnable onCancel) {
+        if (this == CancellationToken.NONE) {
+            return;
+        }
+        OnCancelNode node = new OnCancelNode(onCancel);
+        final boolean added = this.tryAddOnCancelNode(node);
+        if (!added) {
+            // Couldn't add since the list was frozen as a result of cancellation,
+            // invoke this onCancel Callback immediately.
+            node.invokeOnCancel();
+        }
+    }
+
+    /**
+     * Registers the onCancel Callback that will be called when this CancellationToken is canceled.
+     *
+     * This operation of registering the CallBack is non-blocking and thread-safe. If the token
+     * is already in the cancelled state then the provided CallBack will be immediately executed.
+     *
+     * This is O(1) in best case and O(k) in worst, where k is number of concurrent threads in case
+     * of race in registering Callback.
+     *
+     * @param id The registration id for the onCancel Callback.
+     * @param onCancel The onCancel Callback.
+     */
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    public void registerOnCancel(@NonNull String id, @NonNull Runnable onCancel) {
+        if (this == CancellationToken.NONE) {
+            return;
+        }
+        OnCancelNode node = new OnCancelNode(id, onCancel);
+        final boolean added = this.tryAddOnCancelNode(node);
+        if (!added) {
+            // Couldn't add since the list was frozen as a result of cancellation,
+            // invoke this onCancel Callback immediately.
+            node.invokeOnCancel();
+        }
+    }
+
+    /**
+     * Unregister the onCancel Callback.
+     *
+     * This operation of unregistering the CallBack is non-blocking and thread-safe.
+     *
+     * @param id The id of the onCancel Callback to unregister.
+     */
+    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+    public void unregisterOnCancel(@NonNull String id) {
+        if (this == CancellationToken.NONE) {
+            return;
+        }
+        OnCancelNode itr = this.onCancelNodes;
+        // Step_1: Locate the node and mark it as logically deleted.
+        //
+        while (itr != null) {
+            if (this.onCancelNodes == OnCancelNode.FREEZED) {
+                // The token is cancelled, no use in progressing.
+                return;
+            }
+            if (itr.id != null && itr.id.equals(id)) {
+                // Mark the node as logically deleted.
+                itr.markDeleted();
+                break;
+            }
+        }
+        if (itr == null) {
```

Comment:
Since we don't change what `itr` refers to, maybe we can put this before the while loop just to improve readability :)

Reply:
actually there was a bug in the code, which I just fixed.